diff --git a/ldso/dynlink.c b/ldso/dynlink.c
index 6c5cfeb7..1cc11276 100644
--- a/ldso/dynlink.c
+++ b/ldso/dynlink.c
@@ -1117,7 +1117,7 @@ static struct dso *load_library(const char *name, struct dso *needed_by)
 	/* Catch and block attempts to reload the implementation itself */
 	if (name[0]=='l' && name[1]=='i' && name[2]=='b') {
 		static const char reserved[] =
-			"c.pthread.rt.m.dl.util.xnet.";
+			"c.c_musl.pthread.rt.m.dl.util.xnet.";
 		const char *rp, *next;
 		for (rp=reserved; *rp; rp=next) {
 			next = strchr(rp, '.') + 1;
